modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/stat/IgniteStatisticsConfigurationManager.java

```diff
@@ -0,0 +1,549 @@
+        }
+    }
+
+    /** */
+    private void onChangeStatisticConfiguration(
+        StatisticsObjectConfiguration oldCfg,
+        final StatisticsObjectConfiguration newCfg
+    ) {
+        synchronized (mux) {
+            if (log.isDebugEnabled())
+                log.debug("Statistic configuration changed [old=" + oldCfg + ", new=" + newCfg + ']');
+
+            StatisticsObjectConfiguration.Diff diff = StatisticsObjectConfiguration.diff(oldCfg, newCfg);
+
+            if (!F.isEmpty(diff.dropCols()))
+                dropColumnsOnLocalStatistics(newCfg, diff.dropCols());
+
+            if (!F.isEmpty(diff.updateCols())) {
+                GridH2Table tbl = schemaMgr.dataTable(newCfg.key().schema(), newCfg.key().obj());
+
+                GridCacheContext cctx = tbl.cacheContext();
+
+                Set<Integer> parts = cctx.affinity().primaryPartitions(
+                    cctx.localNodeId(), cctx.affinity().affinityTopologyVersion());
+
+                gatherLocalStatistics(
+                    newCfg,
+                    tbl,
+                    parts,
+                    parts,
+                    diff.updateCols()
+                );
+            }
+        }
+    }
+
+    /** */
+    private void gatherLocalStatistics(
+        StatisticsObjectConfiguration cfg,
+        GridH2Table tbl,
+        Set<Integer> partsToAggregate,
+        Set<Integer> partsToCollect,
+        Map<String, StatisticsColumnConfiguration> colsToCollect
+    ) {
+        if (F.isEmpty(colsToCollect))
+            colsToCollect = cfg.columns();
+
+        Column[] cols = IgniteStatisticsHelper.filterColumns(tbl.getColumns(), colsToCollect.keySet());
+
+        gatherer.gatherLocalObjectsStatisticsAsync(
+            tbl,
+            cols,
+            colsToCollect,
+            partsToCollect
+        );
+
+        gatherer.aggregateStatisticsAsync(cfg.key(), () -> aggregateLocalGathering(cfg.key(), partsToAggregate));
+    }
+
+    /** */
+    private void dropColumnsOnLocalStatistics(StatisticsObjectConfiguration cfg, Set<String> cols) {
+        if (log.isDebugEnabled()) {
+            log.debug("Remove local statistics [key=" + cfg.key() +
+                ", columns=" + cols + ']');
+        }
+
+        LocalStatisticsGatheringContext gCtx = gatherer.gatheringInProgress(cfg.key());
+
+        if (gCtx != null) {
+            gCtx.futureAggregate().thenAccept((v) -> {
+                repo.clearLocalStatistics(cfg.key(), cols);
+                repo.clearLocalPartitionsStatistics(cfg.key(), cols);
+            });
+        }
+        else {
+            repo.clearLocalStatistics(cfg.key(), cols);
+            repo.clearLocalPartitionsStatistics(cfg.key(), cols);
+        }
+    }
+
+    /** */
+    private void onDropColumns(GridH2Table tbl, List<String> cols) {
+        assert !F.isEmpty(cols);
+
+        dropStatistics(Collections.singletonList(
+            new StatisticsTarget(
+                tbl.getSchema().getName(),
```

Comment:
Can it differ from tbl.identifier().schema()? Because SchemaManager::dataTable will get table by it's identifier.